generator_process/actions/image_to_image.py

```diff
@@ -0,0 +1,238 @@
+    self,
+    pipeline: Pipeline,
+    
+    model: str,
+
+    scheduler: Scheduler,
+
+    optimizations: Optimizations,
+
+    image: NDArray | str,
+    fit: bool,
+    strength: float,
+    prompt: str,
+    steps: int,
+    width: int,
+    height: int,
+    seed: int,
+
+    cfg_scale: float,
+    use_negative_prompt: bool,
+    negative_prompt: str,
+    
+    seamless: bool,
+    seamless_axes: list[str],
+
+    iterations: int,
+
+    step_preview_mode: StepPreviewMode,
+
+    **kwargs
+) -> Generator[NDArray, None, None]:
+    match pipeline:
+        case Pipeline.STABLE_DIFFUSION:
+            import diffusers
+            import torch
+            from PIL import Image, ImageOps
+            import PIL.Image
+            from ...absolute_path import WEIGHTS_PATH
+
+            # Mostly copied from `diffusers.StableDiffusionImg2ImgPipeline`, with slight modifications to yield the latents at each step.
+            class GeneratorPipeline(diffusers.StableDiffusionImg2ImgPipeline):
+                @torch.no_grad()
+                def __call__(
+                    self,
+                    prompt: Union[str, List[str]],
+                    image: Union[torch.FloatTensor, PIL.Image.Image],
+                    strength: float = 0.8,
+                    num_inference_steps: Optional[int] = 50,
+                    guidance_scale: Optional[float] = 7.5,
+                    negative_prompt: Optional[Union[str, List[str]]] = None,
+                    num_images_per_prompt: Optional[int] = 1,
+                    eta: Optional[float] = 0.0,
+                    generator: Optional[torch.Generator] = None,
+                    output_type: Optional[str] = "pil",
+                    return_dict: bool = True,
+                    callback: Optional[Callable[[int, int, torch.FloatTensor], None]] = None,
+                    callback_steps: Optional[int] = 1,
+                    **kwargs,
+                ):
+                    image = init_image or image
+
+                    # 1. Check inputs
+                    self.check_inputs(prompt, strength, callback_steps)
+
+                    # 2. Define call parameters
+                    batch_size = 1 if isinstance(prompt, str) else len(prompt)
+                    device = self._execution_device
+                    # here `guidance_scale` is defined analog to the guidance weight `w` of equation (2)
+                    # of the Imagen paper: https://arxiv.org/pdf/2205.11487.pdf . `guidance_scale = 1`
+                    # corresponds to doing no classifier free guidance.
+                    do_classifier_free_guidance = guidance_scale > 1.0
+
+                    # 3. Encode input prompt
+                    text_embeddings = self._encode_prompt(
+                        prompt, device, num_images_per_prompt, do_classifier_free_guidance, negative_prompt
+                    )
+
+                    # 4. Preprocess image
+                    if isinstance(image, PIL.Image.Image):
```

Comment:
we very recently had some changes in `diffusers` to allow for batched image input - might be worth slightly adapting here :-)